### PR TITLE
[Contract] Feat/contract claim neg rep

### LINKED ISSUE #378: [UST-166][Contract] Claim Negative Reputation
**Is your feature request related to a problem? Please describe.**
Implement a new function to claim negative reputation after the report is being agreed.

**Describe the solution you'd like**
This function should receive transaction to attest negative reputation to reported person or reporter accordingly.

**Related Issues**
- Claim Negative Reputation (Circuit)
- Claim Positive Reputation (Contract)

**Todos**
- [ ] Implement function `claimReportRep` 

**Tests**
- Should success claim negative reputation
- Should revert with used proof
- Should revert if Epks doesn't come from same user
- Should revert if the postId or commentId doesn't match

### PR DESCRIPTION
**Pull request by FoodChain1028**
## Summary

Implemented Claim Negative Reputation in contract.

## Linked Issue

close #378 

## Details

1. The different type of punishment was not defined in the circuit, so this PR also added the 2 types for negative reputation
    * `type 0`: punishing poster (-5)
    * `type 1`: punishing reporter (-1)
2. Another different design from previous design is the `hash_user_id` is changed into `identity_secret`, which was how previous epochkey get the secret, this is to make them consistent.

## Impacted Areas

**The circuit input is like**:

```typescript
    reportedEpochKey: any
    identitySecret: string | bigint | Identity
    reportedEpoch: number | bigint
    currentEpoch: number | bigint
    currentNonce: number | bigint
    chainId: number | bigint
    attesterId: number | bigint
    type: number | bigint
```
## Tests

- [x] 1. succeed with type 0 (punishing poster)
- [X] 2. revert with used type-0 proof
- [X] 3. succeed with type 1 (punishing reporter)
- [X] 4. revert with used type-1 proof
- [X] 5. revert with invalid proof
- [X] 6. revert with wrong epoch
- [X] 7. revert with invalid attesterId

## Verification Steps

```sh
yarn contracts test
```

## Todo

Break down the progress of the PR for everyone to see what else you intend to include in this PR.

-   [X] Draft implementation for `function`
-   [X] Unit tests for `function`

## Checklist

-   [X] All new and existing tests pass
-   [X] I have updated the documentation (if applicable)
-   [X] My changes do not introduce new security vulnerabilities

### CODE REVIEW

File: packages/contracts/test/ClaimReportNegRep.test.ts [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+        // elapsing 5 epoch
+        await ethers.provider.send('evm_increaseTime', [epochLength * 5])
+        await ethers.provider.send('evm_mine', [])
+
+        const currentNonce = 0
+        const currentEpoch = await posterState.sync.loadCurrentEpoch()
+        expect(currentEpoch).to.be.equal(5)
+
+        const circuit = 'reportNegRepProof'
```
Comment (chentihe) [inline packages/contracts/test/ClaimReportNegRep.test.ts:134]: should use global variable instead (line 32)

File: packages/contracts/test/ClaimReportNegRep.test.ts [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+            identifier,
+            posterPunishment
+        )
+        await expect(tx)
+            .to.emit(app, 'ClaimNegRep')
+            .withArgs(publicSignals[0], currentEpoch)
+
+        posterState.stop()
+    })
+    it('should revert with used proof (for poster punishment)', async () => {
```
Comment (chentihe) [inline packages/contracts/test/ClaimReportNegRep.test.ts:168]: should adding a blank line between tests

File: packages/contracts/test/ClaimReportNegRep.test.ts [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+        // elapsing 5 epoch
+        await ethers.provider.send('evm_increaseTime', [epochLength * 5])
+        await ethers.provider.send('evm_mine', [])
+
+        const currentNonce = 0
+        const currentEpoch = await reporterState.sync.loadCurrentEpoch()
+        expect(currentEpoch).to.be.equal(10)
+
+        const circuit = 'reportNegRepProof'
```
Comment (chentihe) [inline packages/contracts/test/ClaimReportNegRep.test.ts:190]: should use global variable instead (line 32)

File: packages/contracts/test/ClaimReportNegRep.test.ts [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+        // elapsing 5 epoch
+        await ethers.provider.send('evm_increaseTime', [epochLength * 5])
+        await ethers.provider.send('evm_mine', [])
+
+        const currentNonce = 0
+        const currentEpoch = await posterState.sync.loadCurrentEpoch()
+        expect(currentEpoch).to.be.equal(15)
+
+        const circuit = 'reportNegRepProof'
```
Comment (chentihe) [inline packages/contracts/test/ClaimReportNegRep.test.ts:246]: should use global variable instead (line 32)

File: packages/contracts/test/ClaimReportNegRep.test.ts [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+            currentEpoch,
+            currentNonce,
+            chainId,
+            attesterId: wrongAttester,
+            type,
+        })
+
+        // this will be
+        try {
+            await genProofAndVerify(circuit, reportNegRepCircuitInputs)
```
Comment (chentihe) [inline packages/contracts/test/ClaimReportNegRep.test.ts:351]: this function only get { isValid, publicSignals, proof }, should call `claimReportNegRep` to get reverted message.
```typescript
await expect(
    app.claimReportPosRep(
        publicSignals,
        flattenedProof,
        identifier,
         posReputation
    )
).to.be.reverted
```

Comment (FoodChain1028) [reply packages/contracts/test/ClaimReportNegRep.test.ts:351]: This wouldn't correctly get a proof from `genProofAndVerify` :(, so wouldn't get the publicSignals and proof to send in `claimReportPosRep()`

**Review by chentihe — APPROVED** [review]

**Pull request merged**